File: cpp/dnd-character/dnd_character.cpp
#include "dnd_character.h"
#include <algorithm>
#include <array>
#include <cmath>
#include <random>

namespace dnd_character {

// Calculate ability modifier from ability score
int modifier(int score) {
    return static_cast<int>(std::floor((score - 10) / 2.0));
}

// Roll ability score (4d6, drop lowest)
int ability() {
    // Create random number generator
    static std::random_device rd;
    static std::mt19937 gen(rd());
    static std::uniform_int_distribution<> die(1, 6);
    
    // Roll 4 dice
    std::array<int, 4> rolls = {die(gen), die(gen), die(gen), die(gen)};
    
    // Sort the rolls to find the lowest
    std::sort(rolls.begin(), rolls.end());
    
    // Sum the highest 3 rolls (discard the lowest)
    return rolls[1] + rolls[2] + rolls[3];
}

// Constructor to generate a random character
Character::Character() {
    // Generate ability scores
    strength = ability();
    dexterity = ability();
    constitution = ability();
    intelligence = ability();
    wisdom = ability();
    charisma = ability();
    
    // Calculate hitpoints
    hitpoints = 10 + modifier(constitution);
}

}  // namespace dnd_character

File: go/dnd-character/dnd_character.go
package dndcharacter

type Character struct {
	Strength     int
	Dexterity    int
	Constitution int
	Intelligence int
	Wisdom       int
	Charisma     int
	Hitpoints    int
}

// Modifier calculates the ability modifier for a given ability score
func Modifier(score int) int {
	return (score - 10) / 2
}

// Ability uses randomness to generate the score for an ability
func Ability() int {
	// Roll four 6-sided dice
	rolls := make([]int, 4)
	for i := 0; i < 4; i++ {
		rolls[i] = rand.Intn(6) + 1 // 1-6
	}
	
	// Find the smallest roll to discard
	minIndex := 0
	for i := 1; i < 4; i++ {
		if rolls[i] < rolls[minIndex] {
			minIndex = i
		}
	}
	
	// Sum the three highest dice
	sum := 0
	for i, roll := range rolls {
		if i != minIndex {
			sum += roll
		}
	}
	
	return sum
}

// GenerateCharacter creates a new Character with random scores for abilities
func GenerateCharacter() Character {
	character := Character{
		Strength:     Ability(),
		Dexterity:    Ability(),
		Constitution: Ability(),
		Intelligence: Ability(),
		Wisdom:       Ability(),
		Charisma:     Ability(),
	}
	
	// Calculate hitpoints: 10 + constitution modifier
	character.Hitpoints = 10 + Modifier(character.Constitution)
	
	return character
}
